api/lib/domain/models/CampaignParticipationResult.js

```diff
@@ -68,6 +68,16 @@ class CampaignParticipationResult {
   get progress() {
     return campaignParticipationService.progress(this.isCompleted, this.knowledgeElementsCount, this.totalSkillsCount);
   }
+
+  filterPartnerCompetenceResultsWithBadge(badge) {
+    if (badge) {
+      this.partnerCompetenceResults = this.partnerCompetenceResults
+        .filter((partnerCompetenceResult) => partnerCompetenceResult.badgeId === badge.id);
+    }
+    else {
+      this.partnerCompetenceResults.length = 0;
```

Comment:
je suis surpris qu'on affecte la `length`
je pense que s'il n'y a pas de badge, le filter va retourner un tableau vide dans tous les cas.

Reply:
A l'appel de la fonction, `partnerCompetenceResults ` contient tous les `partnerCompetenceResults` de tous les badges. La condition `if` permet de filtrer sur ceux d'un badge spécifique.
Dans le cas contraire, on veut renvoyer un tableau vide au front, d'où le `.length = 0`.

Reply:
Je serais + tenté de faire : `this.partnerCompetenceResults = []` pour bien voir que l'on renvoie un tableau vide

Reply:
Plus généralement, j'ai peur que cette méthode `filterPartnerCompetenceResultsWithBadge` ne soit pas une bonne idée sur son principe. Si j'ai bien compris ce qui se passe, quand on appelle cette méthode, la propriété `partnerCompetenceResults` de l'objet `CampaignParticipationResult` est modifiée pour ne plus contenir que les compétences associées à un certain badge.

Malheureusement, l'objet qui résulte de cette opération n'est plus cohérent : en effet, il possède toujours une liste de `badges` mais pas tous les `partnerCompetenceResults` associés. Plus embêtant, il n'y a aucun moyen de distinguer un objet `CampaignParticipationResult` qui a subi ce traitement, d'un qui reflèterait fidèlement l'état de la base à un moment donné. Autrement dit, une fonction qui recevrait un objet `CampaignParticipationResult` ne peut plus faire de supposition sur les données qu'il contient. De même, un deuxième appel à `filterPartnerCompetenceResultsWithBadge` n'aurait certainement pas l'effet escompté (mais il est impossible de savoir si un premier a été fait).

En fait, l'objet `CampaignParticipationResult` après l'appel de `filterPartnerCompetenceResultsWithBadge` n'est plus un `CampaignParticipationResult` mais une _projection_ de celui-ci sur un badge en particulier. Il devrait donc avoir un type différent pour éviter les erreurs.

Plus généralement, la mutation d'objets est connue comme une source fréquente de bugs et les solutions alternatives existent. Ici par exemple, on pourrait envisager une méthode `getPartnerCompetenceResultsForBadge` qui ferait le filtrage de façon non destructive ? Je vois 
d'ailleurs que la propriété `partnerCompetenceResults` est elle-même une concaténation des résultats par badge, est-ce qu'on ne gagnerait pas à les conserver dans des listes séparées par badge ?

Reply:
Suite à la remarque de Jonathan : 
Ne faudrait-il pas ajouter un tableau `visiblePartnerCompetenceResults` qui serait vide de base et remplit quand on veut afficher les compétences des partenaires ?

Reply:
@MelanieMEB le cœur de ma remarque c'était quand même d'arrêter de modifier l'objet, donc un tableau vide de base et qui se remplit ça ne va pas vraiment dans ce sens ?

Reply:
Ce n'est pas parfait, mais ça évite de modifier le `partnerCompetenceResults` après sa création. Un nouvel attribut sur les CampaignParticipationResult pour afficher, s'il y en a, des partner Competence qui représente les résultats, ne règle pas tout. 
Est-ce qu'une méthode sur l'objet comme `getPartnerCompetenceResultsForBadge` suffirait tu penses pour éviter les problèmes de tableaux remplis/vide selon les besoins ?

Reply:
Oui, c'est pour ça que je proposais une méthode qui prend le badge et renvoie les résultats associés au badge (`getPartnerCompetenceResultsForBadge`), ainsi l'état de l'objet n'est jamais modifié après sa création. En revanche si le `CampaignParticipationResult` doit être sérialisé y compris avec le détail des compétences par badge, alors une méthode ne suffira pas ; dans ce cas je suggère de sérialiser la map badge -> compétences pour tous les badges présents.